File: projects/argo-energy-solutions/backend/python_scripts/ingest/ingest_to_postgres.py
#!/usr/bin/env python3
"""
Eniscope Data Ingestion to PostgreSQL (Neon)

Pulls data from Eniscope API and stores it in PostgreSQL database.

Usage:
    python ingest_to_postgres.py --site 23271 --days 90
"""

import os
import sys
import argparse
import hashlib
import base64
import time
from pathlib import Path
from datetime import datetime, timedelta
from typing import List, Dict, Optional
import requests
import psycopg2
from psycopg2.extras import execute_batch
from dotenv import load_dotenv

_PKG_ROOT = Path(__file__).resolve().parent.parent
_PROJECT_ROOT = _PKG_ROOT.parent.parent
if str(_PKG_ROOT) not in sys.path:
    sys.path.insert(0, str(_PKG_ROOT))
load_dotenv(_PROJECT_ROOT / '.env')


class EniscopeClient:
    """Client for Eniscope API with authentication and rate limiting."""
    
    def __init__(self):
        self.base_url = os.getenv('VITE_ENISCOPE_API_URL', 'https://core.eniscope.com').rstrip('/')
        self.api_key = os.getenv('VITE_ENISCOPE_API_KEY')
        self.email = os.getenv('VITE_ENISCOPE_EMAIL')
        self.password = os.getenv('VITE_ENISCOPE_PASSWORD')
        
        if not all([self.api_key, self.email, self.password]):
            raise ValueError('Missing required environment variables')
        
        self.password_md5 = hashlib.md5(self.password.encode()).hexdigest()
        self.session_token = None
        self.cached_organizations = None
    
    def authenticate(self) -> List[Dict]:
        """Authenticate and return organizations list."""
        if self.cached_organizations:
            return self.cached_organizations
        
        auth_string = f"{self.email}:{self.password_md5}"
        auth_b64 = base64.b64encode(auth_string.encode()).decode()
        
        headers = {
            'Authorization': f'Basic {auth_b64}',
            'X-Eniscope-API': self.api_key,
            'Accept': 'text/json'
        }
        
        response = requests.get(f'{self.base_url}/organizations', headers=headers)
        response.raise_for_status()
        
        self.session_token = response.headers.get('x-eniscope-token') or response.headers.get('X-Eniscope-Token')
        self.cached_organizations = response.json()
        
        return self.cached_organizations
    
    def _make_request_with_retry(self, url: str, params: Dict = None, retries: int = 3) -> requests.Response:
        """Make request with exponential backoff retry."""
        headers = {
            'X-Eniscope-API': self.api_key,
            'X-Eniscope-Token': self.session_token,
            'Accept': 'text/json'
        }
        
        for attempt in range(retries):
            try:
                response = requests.get(url, params=params, headers=headers)
                response.raise_for_status()
                return response
            
            except requests.exceptions.HTTPError as e:
                if e.response.status_code == 429 and attempt < retries - 1:
                    delay = 2 ** (attempt + 3)  # 8s, 16s, 32s
                    print(f"\n   Rate limited. Waiting {delay}s before retry...")
                    time.sleep(delay)
                elif e.response.status_code in (401, 419) and attempt < retries - 1:
                    self.session_token = None
                    self.authenticate()
                else:
                    raise
        
        raise Exception(f"Failed after {retries} retries")
    
    def get_channels(self, organization_id: str) -> List[Dict]:
        """Get channels for an organization."""
        response = self._make_request_with_retry(
            f'{self.base_url}/channels',
            params={'organization': organization_id}
        )
        
        data = response.json()
        
        # Handle various response formats
        if isinstance(data, list):
            return data
        elif isinstance(data, dict):
            return data.get('channels') or data.get('data') or data.get('items') or []
        return []
    
    def get_readings(self, channel_id: int, start_date: str, end_date: str, 
                     fields: List[str] = None, resolution: int = 900) -> List[Dict]:
        """Get readings for a channel."""
        if fields is None:
            fields = ['E', 'P', 'V', 'I', 'PF']
        
        # Convert dates to Unix timestamps
        start_ts = int(datetime.fromisoformat(start_date.replace('Z', '+00:00')).timestamp())
        end_ts = int(datetime.fromisoformat(end_date.replace('Z', '+00:00')).timestamp())
        
        params = {
            'action': 'summarize',
            'res': str(resolution),
            'daterange[]': [start_ts, end_ts],
            'fields[]': fields
        }
        
        response = self._make_request_with_retry(
            f'{self.base_url}/readings/{channel_id}',
            params=params
        )
        
        data = response.json()
        
        # Extract readings array
        readings = []
        if isinstance(data, list):
            readings = data
        elif isinstance(data, dict):
            readings = (data.get('records') or data.get('data') or 
                       data.get('result') or data.get('readings') or [])
        
        # Normalize and convert units (Wh -> kWh, W -> kW)
        normalized = []
        for r in readings:
            normalized.append({
                'timestamp': r.get('ts') or r.get('t') or r.get('timestamp'),
                'energy_kwh': r.get('E') / 1000 if r.get('E') is not None else None,
                'power_kw': r.get('P') / 1000 if r.get('P') is not None else None,
                'voltage_v': r.get('V'),
                'current_a': r.get('I'),
                'power_factor': r.get('PF')
            })
        
        return normalized


class PostgresDB:
    """PostgreSQL database operations."""
    
    def __init__(self, connection_string: str):
        self.conn = psycopg2.connect(connection_string)
        self.conn.autocommit = False
    
    def __enter__(self):
        return self
    
    def __exit__(self, exc_type, exc_val, exc_tb):
        if exc_type:
            self.conn.rollback()
        else:
            self.conn.commit()
        self.conn.close()
    
    def upsert_organization(self, org_id: str, org_name: str):
        """Insert or update organization."""
        with self.conn.cursor() as cur:
            cur.execute("""
                INSERT INTO organizations (organization_id, organization_name, updated_at)
                VALUES (%s, %s, NOW())
                ON CONFLICT (organization_id) 
                DO UPDATE SET organization_name = EXCLUDED.organization_name, updated_at = NOW()
            """, (org_id, org_name))
        self.conn.commit()
    
    def upsert_device(self, device_id: int, device_name: str, device_type: str, 
                     uuid: str, org_id: str) -> bool:
        """Insert or update device."""
        try:
            with self.conn.cursor() as cur:
                cur.execute("""
                    INSERT INTO devices (device_id, device_name, device_type, 
                                       serial_number, organization_id, updated_at)
                    VALUES (%s, %s, %s, %s, %s, NOW())
                    ON CONFLICT (device_id)
                    DO UPDATE SET 
                        device_name = EXCLUDED.device_name,
                        device_type = EXCLUDED.device_type,
                        serial_number = EXCLUDED.serial_number,
                        updated_at = NOW()
                """, (device_id, device_name, device_type, uuid, org_id))
            self.conn.commit()
            return True
        except Exception as e:
            print(f"   ⚠️  Error storing device {device_id}: {e}")
            self.conn.rollback()
            return False
    
    def upsert_channel(self, channel_id: int, channel_name: str, org_id: str, 
                      device_id: int = None) -> bool:
        """Insert or update channel."""
        try:
            with self.conn.cursor() as cur:
                cur.execute("""
                    INSERT INTO channels (channel_id, channel_name, organization_id, 
                                        device_id, channel_type, unit, updated_at)
                    VALUES (%s, %s, %s, %s, %s, %s, NOW())
                    ON CONFLICT (channel_id)
                    DO UPDATE SET 
                        channel_name = EXCLUDED.channel_name,
                        device_id = EXCLUDED.device_id,
                        updated_at = NOW()
                """, (channel_id, channel_name, org_id, device_id, 'energy', 'kWh'))
            self.conn.commit()
            return True
        except Exception as e:
            print(f"   ⚠️  Error storing channel {channel_id}: {e}")
            self.conn.rollback()
            return False
    
    def insert_readings(self, channel_id: int, readings: List[Dict]) -> int:
        """Batch insert readings."""
        if not readings:
            return 0
        
        inserted = 0
        batch_size = 1000
        
        with self.conn.cursor() as cur:
            for i in range(0, len(readings), batch_size):
                batch = readings[i:i + batch_size]
                
                # Filter out readings with null timestamps
                valid_batch = [
                    (channel_id,
                     datetime.fromtimestamp(r['timestamp']) if isinstance(r['timestamp'], (int, float)) 
                     else datetime.fromisoformat(str(r['timestamp']).replace('Z', '+00:00')),
                     r['energy_kwh'],
                     r['power_kw'],
                     r['voltage_v'],
                     r['current_a'],
                     r['power_factor'])
                    for r in batch
                    if r.get('timestamp') is not None
                ]
                
                if not valid_batch:
                    continue
                
                try:
                    execute_batch(cur, """
                        INSERT INTO readings (channel_id, timestamp, energy_kwh, power_kw,
                                            voltage_v, current_a, power_factor)
                        VALUES (%s, %s, %s, %s, %s, %s, %s)
                        ON CONFLICT (channel_id, timestamp) DO NOTHING
                    """, valid_batch)
                    
                    inserted += cur.rowcount
                except Exception as e:
                    print(f"   Error inserting batch: {e}")
                    self.conn.rollback()
                    continue
        
        self.conn.commit()
        return inserted
    
    def get_total_readings(self) -> int:
        """Get total number of readings in database."""
        with self.conn.cursor() as cur:
            cur.execute("SELECT COUNT(*) FROM readings")
            return cur.fetchone()[0]


def ingest_data(site_id: str, days: int):
    """Main ingestion function."""
    print("🌐 Eniscope → PostgreSQL Data Ingestion\n")
    print(f"📊 Site ID: {site_id}")
    print(f"📅 Days to fetch: {days}\n")
    
    # Check environment
    db_url = os.getenv('DATABASE_URL')
    if not db_url:
        print("❌ DATABASE_URL not found in .env")
        sys.exit(1)
    
    # Initialize clients
    client = EniscopeClient()
    
    try:
        # Authenticate
        orgs = client.authenticate()
        print("✅ Authenticated with Eniscope\n")
        
        # Get organization
        print("📋 Fetching organization...")
        org = None
        if isinstance(orgs, list):
            org = next((o for o in orgs if str(o.get('organization_id') or o.get('id')) == site_id), 
                      orgs[0] if orgs else None)
        elif isinstance(orgs, dict):
            orgs_list = orgs.get('organizations') or orgs.get('data') or [orgs]
            org = next((o for o in orgs_list if str(o.get('organization_id') or o.get('id')) == site_id), 
                      orgs_list[0] if orgs_list else None)
        
        if not org:
            print(f"❌ Organization {site_id} not found")
            sys.exit(1)
        
        org_name = org.get('organization_name') or org.get('name') or f"Site {site_id}"
        print(f"✅ Organization: {org_name}\n")
        
        # Connect to database
        with PostgresDB(db_url) as db:
            print("✅ Connected to PostgreSQL\n")
            
            # Store organization
            db.upsert_organization(site_id, org_name)
            
            # Get channels
            print("🔌 Fetching channels...")
            channels = client.get_channels(site_id)
            print(f"✅ Found {len(channels)} channels\n")
            
            # Store channels and devices
            valid_channels = []
            for channel in channels:
                channel_id = channel.get('channelId') or channel.get('dataChannelId')
                channel_name = channel.get('channelName') or channel.get('name') or f"Channel {channel_id}"
                
                if not channel_id:
                    continue
                
                # Extract device information
                device_id = channel.get('deviceId')
                device_name = channel.get('deviceName')
                device_type = channel.get('deviceTypeName') or channel.get('deviceType')
                uuid = channel.get('uuId') or channel.get('uuid')
                
                # Store device if present
                if device_id:
                    db.upsert_device(
                        int(device_id),
                        device_name or f"Device {device_id}",
                        device_type or 'Unknown',
                        uuid or '',
                        site_id
                    )
                
                # Store channel with device link
                if db.upsert_channel(int(channel_id), channel_name, site_id, 
                                    int(device_id) if device_id else None):
                    valid_channels.append({
                        'id': int(channel_id),
                        'name': channel_name
                    })
            
            print(f"✅ {len(valid_channels)} valid channels stored\n")
            
            # Calculate date range
            end_date = datetime.now()
            start_date = end_date - timedelta(days=days)
            date_range = f"{start_date.date()}/{end_date.date()}"
            
            print(f"📅 Date range: {date_range}\n")
            print("📥 Fetching readings...\n")
            
            # Fetch and store readings
            total_readings = 0
            start_time = time.time()
            
            for i, channel in enumerate(valid_channels, 1):
                print(f"   [{i}/{len(valid_channels)}] {channel['name']}... ", end='', flush=True)
                
                try:
                    readings = client.get_readings(
                        channel['id'],
                        start_date.isoformat(),
                        end_date.isoformat(),
                        fields=['E', 'P', 'V', 'I', 'PF'],
                        resolution=900  # 15 minutes
                    )
                    
                    inserted = db.insert_readings(channel['id'], readings)
                    total_readings += inserted
                    
                    print(f"✅ {inserted:,} readings")
                    
                    # Delay to avoid rate limiting
                    time.sleep(1.5)
                    
                except Exception as e:
                    print(f"❌ {e}")
            
            duration = time.time() - start_time
            
            print(f"\n✅ Ingestion complete!")
            print(f"   Total readings: {total_readings:,}")
            print(f"   Duration: {duration:.1f}s\n")
            
            # Verify
            total_in_db = db.get_total_readings()
            print(f"📊 Total readings in database: {total_in_db:,}\n")
            
            # Refresh materialized views so analytics stay current
            try:
                from govern.refresh_views import refresh_materialized_views
                print("🔄 Refreshing materialized views...")
                if refresh_materialized_views(verbose=True):
                    print("   ✅ Views refreshed.\n")
                else:
                    print("   ⚠️  Some view refreshes skipped or failed.\n")
            except Exception as e:
                print(f"   ⚠️  View refresh skipped: {e}\n")
            
            print("💡 Next steps:")
            print("   1. Enable TimescaleDB: See docs/setup/NEON_SETUP_GUIDE.md")
            print("   2. Set up daily sync: Add to crontab")
            print("   3. Query your data with Python!")
    
    except Exception as e:
        print(f"\n❌ Error: {e}")
        import traceback
        traceback.print_exc()
        sys.exit(1)


def main():
    parser = argparse.ArgumentParser(description='Ingest Eniscope data to PostgreSQL')
    parser.add_argument('--site', default='23271', help='Site ID (default: 23271)')
    parser.add_argument('--days', type=int, default=90, help='Days of data to fetch (default: 90)')
    
    args = parser.parse_args()
    ingest_data(args.site, args.days)


if __name__ == '__main__':
    main()

File: projects/argo-energy-solutions/backend/scripts/data-collection/ingest-to-postgres.js
#!/usr/bin/env node

/**
 * Eniscope Data Ingestion to PostgreSQL (Neon)
 * 
 * Pulls data from Eniscope API and stores it in PostgreSQL database
 * 
 * Usage:
 *   node ingest-to-postgres.js --site=23271 --days=90
 */

import pg from 'pg';
import axios from 'axios';
import crypto from 'crypto';
import dotenv from 'dotenv';
import path from 'path';
import { fileURLToPath } from 'url';

const __filename = fileURLToPath(import.meta.url);
const __dirname = path.dirname(__filename);

dotenv.config({ path: path.join(__dirname, '..', '..', '..', '.env') });

const { Pool } = pg;

// Eniscope API Client
class EniscopeClient {
  constructor() {
    this.baseUrl = (process.env.VITE_ENISCOPE_API_URL || 'https://core.eniscope.com').replace(/\/$/, '');
    this.apiKey = process.env.VITE_ENISCOPE_API_KEY;
    this.email = process.env.VITE_ENISCOPE_EMAIL;
    this.password = process.env.VITE_ENISCOPE_PASSWORD;
    this.passwordMd5 = crypto.createHash('md5').update(this.password).digest('hex');
    this.sessionToken = null;
  }

  async authenticate() {
    if (this.sessionToken) return this.cachedOrganizations || [];

    const authString = `${this.email}:${this.passwordMd5}`;
    const authB64 = Buffer.from(authString).toString('base64');
    
    try {
      const response = await axios.get(`${this.baseUrl}/organizations`, {
        headers: {
          'Authorization': `Basic ${authB64}`,
          'X-Eniscope-API': this.apiKey,
          'Accept': 'text/json'
        }
      });

      this.sessionToken = response.headers['x-eniscope-token'] || response.headers['X-Eniscope-Token'] || null;
      this.cachedOrganizations = response.data; // Cache organizations from auth call
      return this.cachedOrganizations;
    } catch (error) {
      console.error('Authentication failed:', error.response?.data || error.message);
      throw error;
    }
  }

  async getOrganizations() {
    // Return cached orgs from authentication to avoid extra API call
    if (this.cachedOrganizations) {
      return this.cachedOrganizations;
    }
    return await this.authenticate();
  }

  async makeRequestWithRetry(url, config, retries = 3) {
    for (let attempt = 0; attempt < retries; attempt++) {
      try {
        return await axios.get(url, config);
      } catch (error) {
        if (error.response?.status === 429 && attempt < retries - 1) {
          const delay = Math.pow(2, attempt + 3) * 1000; // 8s, 16s, 32s
          console.log(`\n   Rate limited. Waiting ${delay/1000}s before retry...`);
          await new Promise(resolve => setTimeout(resolve, delay));
        } else if ((error.response?.status === 401 || error.response?.status === 419) && attempt < retries - 1) {
          this.sessionToken = null;
          await this.authenticate();
        } else {
          throw error;
        }
      }
    }
  }

  async getChannels(organizationId) {
    await this.authenticate();
    const response = await this.makeRequestWithRetry(`${this.baseUrl}/channels`, {
      params: { organization: organizationId },
      headers: {
        'X-Eniscope-API': this.apiKey,
        'X-Eniscope-Token': this.sessionToken,
        'Accept': 'text/json'
      }
    });
    
    // Handle various response formats
    if (Array.isArray(response.data)) {
      return response.data;
    } else if (response.data?.channels) {
      return response.data.channels;
    } else if (response.data?.data) {
      return response.data.data;
    }
    return [];
  }

  async getReadings(channelId, options = {}) {
    await this.authenticate();
    
    const params = {
      action: 'summarize',
      res: options.resolution || '900' // 15 minutes in seconds
    };
    
    // Add fields
    if (options.fields && Array.isArray(options.fields)) {
      options.fields.forEach(field => {
        if (!params['fields[]']) {
          params['fields[]'] = [];
        }
        params['fields[]'].push(field);
      });
    }
    
    // Add daterange as array
    if (options.daterange) {
      const dates = options.daterange.split('/');
      params['daterange[]'] = dates.map(d => Math.floor(new Date(d).getTime() / 1000));
    }
    
    const response = await this.makeRequestWithRetry(`${this.baseUrl}/readings/${channelId}`, {
      params,
      headers: {
        'X-Eniscope-API': this.apiKey,
        'X-Eniscope-Token': this.sessionToken,
        'Accept': 'text/json'
      }
    });

    let readings = [];
    if (Array.isArray(response.data)) {
      readings = response.data;
    } else if (response.data?.records) {
      readings = response.data.records;
    } else if (response.data?.data) {
      readings = response.data.data;
    } else if (response.data?.result) {
      readings = response.data.result;
    }

    // Skip if no readings
    if (!Array.isArray(readings) || readings.length === 0) {
      return [];
    }

    // Debug: Log first reading to see structure (first 3 channels only)
    if (readings.length > 0) {
      const firstReading = readings[0];
      console.log(`\n   API Response - First reading keys: ${Object.keys(firstReading).join(', ')}`);
    }

    // Convert units: Wh -> kWh, W -> kW
    return readings.map(r => ({
      ...r,
      t: r.ts || r.t || r.timestamp || r.time, // Normalize timestamp field (API uses 'ts')
      E: r.E != null ? r.E / 1000 : null,
      P: r.P != null ? r.P / 1000 : null
    }));
  }
}

// Database functions
async function upsertOrganization(pool, org, orgId) {
  const id = String(orgId || org.organization_id || org.id);
  const name = org.organization_name || org.name || org.organizationname || `Organization ${id}`;
  
  await pool.query(`
    INSERT INTO organizations (
      organization_id, organization_name, address, city, postcode, country, updated_at
    ) VALUES ($1, $2, $3, $4, $5, $6, NOW())
    ON CONFLICT (organization_id) 
    DO UPDATE SET 
      organization_name = EXCLUDED.organization_name,
      updated_at = NOW()
  `, [
    id,
    name,
    org.address || null,
    org.city || null,
    org.postcode || null,
    org.country || null
  ]);
}

async function upsertChannel(pool, channel, organizationId) {
  // Extract channel ID using correct field names
  const channelId = channel.channelId || channel.dataChannelId || channel.channel_id || channel.id;
  const channelName = channel.channelName || channel.channel_name || channel.name || `Channel ${channelId}`;
  const deviceId = channel.deviceId || channel.device_id || channel.device;
  
  // Skip if no valid channel ID
  if (!channelId) {
    console.log(`   ⚠️  Skipping channel with no ID`);
    return false;
  }
  
  try {
    // Don't include device_id to avoid foreign key constraint issues
    await pool.query(`
      INSERT INTO channels (
        channel_id, channel_name, organization_id, channel_type, unit, updated_at
      ) VALUES ($1, $2, $3, $4, $5, NOW())
      ON CONFLICT (channel_id)
      DO UPDATE SET
        channel_name = EXCLUDED.channel_name,
        updated_at = NOW()
    `, [
      parseInt(channelId),
      channelName,
      String(organizationId),
      channel.type || 'energy',
      channel.unit || 'kWh'
    ]);
    
    return true;
  } catch (error) {
    console.log(`   ⚠️  Error storing channel ${channelId}: ${error.message}`);
    return false;
  }
}

async function insertReadings(pool, channelId, readings) {
  if (readings.length === 0) return 0;

  const batchSize = 1000;
  let inserted = 0;

  for (let i = 0; i < readings.length; i += batchSize) {
    const batch = readings.slice(i, Math.min(i + batchSize, readings.length));
    
    const values = batch.map((r, idx) => {
      const offset = idx * 7;
      return `($${offset+1}, $${offset+2}, $${offset+3}, $${offset+4}, $${offset+5}, $${offset+6}, $${offset+7})`;
    }).join(',');

    const params = batch.flatMap(r => {
      // Handle timestamp - could be Unix timestamp (number) or ISO string
      let timestamp;
      if (typeof r.t === 'number') {
        timestamp = new Date(r.t * 1000); // Unix timestamp in seconds
      } else if (typeof r.t === 'string') {
        timestamp = new Date(r.t); // ISO string
      } else if (r.timestamp) {
        timestamp = new Date(r.timestamp);
      } else {
        timestamp = null;
      }
      
      return [
        channelId,
        timestamp,
        r.E || null,
        r.P || null,
        r.V || null,
        r.I || null,
        r.PF || null
      ];
    });

    try {
      const result = await pool.query(`
        INSERT INTO readings (
          channel_id, timestamp, energy_kwh, power_kw, voltage_v, current_a, power_factor
        ) VALUES ${values}
        ON CONFLICT (channel_id, timestamp) DO NOTHING
      `, params);

      inserted += result.rowCount || 0;
    } catch (error) {
      console.error(`Error inserting batch: ${error.message}`);
    }
  }

  return inserted;
}

// Main ingestion function
async function ingestData() {
  console.log('🌐 Eniscope → PostgreSQL Data Ingestion\n');

  // Parse command line arguments
  const args = process.argv.slice(2);
  const siteArg = args.find(arg => arg.startsWith('--site='));
  const daysArg = args.find(arg => arg.startsWith('--days='));

  const siteId = siteArg ? siteArg.split('=')[1] : '23271'; // Default to Wilson Center
  const days = daysArg ? parseInt(daysArg.split('=')[1]) : 90;

  console.log(`📊 Site ID: ${siteId}`);
  console.log(`📅 Days to fetch: ${days}\n`);

  // Check environment
  if (!process.env.DATABASE_URL) {
    console.error('❌ DATABASE_URL not found in .env');
    process.exit(1);
  }

  // Initialize
  const client = new EniscopeClient();
  const pool = new Pool({
    connectionString: process.env.DATABASE_URL,
    ssl: { rejectUnauthorized: false },
    max: 10
  });

  try {
    // Test connection
    await pool.query('SELECT NOW()');
    console.log('✅ Connected to PostgreSQL\n');

    // Authenticate with Eniscope
    await client.authenticate();
    console.log('✅ Authenticated with Eniscope\n');

    // Get organization info
    console.log('📋 Fetching organization...');
    let orgsData = await client.getOrganizations();
    
    // Handle various response formats
    let orgs = [];
    if (Array.isArray(orgsData)) {
      orgs = orgsData;
    } else if (orgsData && typeof orgsData === 'object') {
      orgs = orgsData.organizations || orgsData.data || orgsData.items || [orgsData];
    }
    
    const org = orgs.find(o => (o.organization_id || o.id) == siteId) || orgs[0];
    
    if (!org) {
      throw new Error(`Organization ${siteId} not found`);
    }

    await upsertOrganization(pool, org, siteId);
    console.log(`✅ Organization: ${org.organization_name || org.name || `Site ${siteId}`}\n`);

    // Get channels
    console.log('🔌 Fetching channels...');
    const allChannels = await client.getChannels(siteId);
    console.log(`✅ Found ${allChannels.length} channels\n`);

    // Store channels (filter out invalid ones)
    const validChannels = [];
    for (const channel of allChannels) {
      const success = await upsertChannel(pool, channel, siteId);
      if (success) {
        validChannels.push(channel);
      }
    }
    console.log(`✅ ${validChannels.length} valid channels stored\n`);
    
    // Use only valid channels for data fetching
    const channels = validChannels;

    // Calculate date range
    const endDate = new Date();
    const startDate = new Date();
    startDate.setDate(startDate.getDate() - days);

    const dateRange = `${startDate.toISOString().split('T')[0]}/${endDate.toISOString().split('T')[0]}`;
    
    console.log(`📅 Date range: ${dateRange}\n`);
    console.log('📥 Fetching readings...\n');

    // Fetch and store readings for each channel
    let totalReadings = 0;
    const startTime = Date.now();

    for (let i = 0; i < channels.length; i++) {
      const channel = channels[i];
      const channelId = channel.channelId || channel.dataChannelId || channel.channel_id || channel.id;
      const channelName = channel.channelName || channel.channel_name || channel.name || `Channel ${channelId}`;

      process.stdout.write(`   [${i+1}/${channels.length}] ${channelName}... `);

      try {
        const readings = await client.getReadings(channelId, {
          fields: ['E', 'P', 'V', 'I', 'PF'],
          daterange: dateRange,
          resolution: '900' // 15 minutes in seconds
        });

        const inserted = await insertReadings(pool, channelId, readings);
        totalReadings += inserted;

        console.log(`✅ ${inserted.toLocaleString()} readings`);

        // Delay to avoid rate limiting (1-2 seconds per channel)
        await new Promise(resolve => setTimeout(resolve, 1500));

      } catch (error) {
        console.log(`❌ ${error.message}`);
      }
    }

    const duration = ((Date.now() - startTime) / 1000).toFixed(1);

    console.log(`\n✅ Ingestion complete!`);
    console.log(`   Total readings: ${totalReadings.toLocaleString()}`);
    console.log(`   Duration: ${duration}s\n`);

    // Verify
    const countResult = await pool.query('SELECT COUNT(*) FROM readings');
    console.log(`📊 Total readings in database: ${parseInt(countResult.rows[0].count).toLocaleString()}\n`);

    console.log('💡 Next steps:');
    console.log('   1. Generate report: npm run report:weekly -- --site 23271 --db postgres');
    console.log('   2. Set up daily sync: Add to crontab\n');

  } catch (error) {
    console.error('\n❌ Error:', error.message);
    console.error(error.stack);
    process.exit(1);
  } finally {
    await pool.end();
  }
}

ingestData();
